google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/JdbcDatabaseMetaData.java

```diff
@@ -0,0 +1,1390 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner.jdbc;
+
+import java.io.BufferedReader;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.RowIdLifetime;
+import java.sql.SQLException;
+import java.sql.Types;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Scanner;
+import com.google.cloud.spanner.ResultSets;
+import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.Type;
+import com.google.cloud.spanner.Type.StructField;
+
+/** {@link DatabaseMetaData} implementation for Cloud Spanner */
+class JdbcDatabaseMetaData extends AbstractJdbcWrapper implements DatabaseMetaData {
+  private static final int JDBC_MAJOR_VERSION = 4;
+  private static final int JDBC_MINOR_VERSION = 1;
+  private static final int DATABASE_MAJOR_VERSION = 1;
+  private static final int DATABASE_MINOR_VERSION = 0;
+  private static final String PRODUCT_NAME = "Google Cloud Spanner";
+
+  private String readSqlFromFile(String filename) {
+    InputStream in = getClass().getResourceAsStream(filename);
+    BufferedReader reader = new BufferedReader(new InputStreamReader(in));
+    StringBuilder builder = new StringBuilder();
+    try (Scanner scanner = new Scanner(reader)) {
+      while (scanner.hasNextLine()) {
+        String line = scanner.nextLine();
+        builder.append(line).append("\n");
+      }
+      scanner.close();
+    }
+    return builder.toString();
+  }
+
+  private JdbcConnection connection;
+
+  JdbcDatabaseMetaData(JdbcConnection connection) {
+    this.connection = connection;
+  }
+
+  @Override
+  public boolean isClosed() {
+    return false;
+  }
+
+  @Override
+  public boolean allProceduresAreCallable() throws SQLException {
+    return true;
+  }
+
+  @Override
+  public boolean allTablesAreSelectable() throws SQLException {
+    return true;
+  }
+
+  @Override
+  public String getURL() throws SQLException {
+    return connection.getConnectionUrl();
+  }
+
+  @Override
+  public String getUserName() throws SQLException {
+    // TODO: extract from credentials
```

Comment:
still have a todo here?

Reply:
Added implementation for `getUserName()` for known `Credentials` types.

Reply:
I've left the `getFunctions(String, String, String)` method as is (i.e. returning an empty `ResultSet`), as there is no server side equivalent for this in Cloud Spanner, and this is hardly ever used